JS.md

```diff
@@ -0,0 +1,237 @@
+# JavaScript API
+
+In the [MVP](MVP.md), the only way to access WebAssembly on the Web is through
+an explicit JS API which is defined below. (In the future, WebAssembly may also
+be loaded and run directly from an HTML `<script type='module'>` tag—and
+any other Web API that loads ES6 modules via URL—as part of 
+[ES6 Module integration](Modules.md#integration-with-es6-modules).)
+
+## The `Wasm` object
+
+The `Wasm` object is the initial value of the `Wasm` property of the global
+object. Like the `Math` and `JSON` objects, the `Wasm` object is a plain
+JS object (not a constructor or function) that acts like a namespace and has the
+following properties:
+
+### Constructor Properties of the `Wasm` object
+
+* `Wasm.Module` : see [`Wasm.Module` objects](#wasmmodule-objects)
+* `Wasm.Instance` : see [`Wasm.Instance` objects](#wasminstance-objects)
+* `Wasm.CompileError` : a [NativeError](http://tc39.github.io/ecma262/#sec-nativeerror-object-structure)
+   which indicates an error during WebAssembly decoding or validation
+* `Wasm.RuntimeError` : a [NativeError](http://tc39.github.io/ecma262/#sec-nativeerror-object-structure)
+   which indicates an error while running WebAssembly code
+
+### Function Properties of the `Wasm` object
+
+#### `Wasm.compile`
+
+The `compile` function has the signature:
+```
+Promise<Wasm.Module> compile(TypedArray bytes)
+```
+If the given `bytes` argument is not a
+[TypedArray object](http://tc39.github.io/ecma262/#sec-typedarray-objects),
+the returned `Promise` is [rejected](http://tc39.github.io/ecma262/#sec-rejectpromise)
+with a `TypeError`.
+
+Otherwise, this function starts an asychronous task to compile a `Wasm.Module`
+as described in the [`Wasm.Module` constructor](#wasmmodule-constructor).
+On success, the `Promise` is [fulfilled](http://tc39.github.io/ecma262/#sec-fulfillpromise)
+with the resulting `Wasm.Module` instance. On failure, the `Promise` is 
+[rejected](http://tc39.github.io/ecma262/#sec-rejectpromise) with the 
+`Wasm.CompileError`.
+
+The asynchronous compilation is logically performed on a copy of the state of
+the given TypedArray captured during the call to `compile`; subsequent mutations
+of the TypedArray after `compile` return do not affect ongoing compilations.
+
+In the [future](FutureFeatures.md#streaming-compilation), this function can be
+extended to accept a [stream](https://streams.spec.whatwg.org), thereby enabling
+asynchronous, background, streaming compilation.
+
+## `Wasm.Module` Objects
```

Comment:
/me goes to see what a Module Namespace Exotic Object is...

I think you mean `Wasm.Instance`; a `Wasm.Module` doesn't have a direct correspondence to anything in ES6, afaics.

So for `Instance`, that is a good question and I like the symmetry.  We had wanted to avoid injecting all the export names directly into the instance object since these could shadow prototype methods.  Looks like Module Namespace Objects don't have that problem; there is not even a prototype.  There is currently only one method on `Wasm.Instance.prototype`: `instantiate`, so I see two options that avoid accidental shadowing:
* make the `exports` field of `Wasm.Instance` objects hold a Module Namespace Exotic Object
* convert `instantiate` into the `Wasm.Instance` constructor

I had considered the latter when initially writing this PR, but I thought that `m.instantiate()` looked a bit nicer.  But maybe it's more explicit to see the `new` and know that there are two distinct objects here.  WDYT?